File: src/template/binary_indexed_tree/Fenwick.java
package template.binary_indexed_tree;

/**
 * @author chenqwwq
 * @date 2022/10/7
 **/
public class Fenwick {

    private final int[] tree;
    private final int cap;


    public Fenwick(int cap) {
        this.cap = cap;
        tree = new int[cap];
    }

    public Fenwick(int[] source) {
        this.cap = source.length;
        this.tree = new int[cap];
        System.arraycopy(source, 0, tree, 0, cap);

    }

    public void add(int p, int num) {
        for (; p < cap; p += lowbit(p)) {
            tree[p] += num;
        }
    }

    public long query(int p) {
        long ans = 0;
        for (; p > 0; p -= lowbit(p)) {
            ans += tree[p];
        }
        return ans;
    }

    private int lowbit(int x) {
        return x & -x;
    }

}

File: src/cpp/template/binary_indexed_tree/fenwick.cpp
//
// Created by 陈炳鑫 on 2022/4/4.
//


#include "bits/stdc++.h"
#include "common.h"

using namespace std;

/**
 * 树状数组
 */
class fenwick {
private:
    vector<int> tree;
    int n;

    int lowbit(int i) {
        return i & -i;
    }

public:
    fenwick(vector<int> &ds) {
        n = (int) ds.size();
        tree.resize(n + 1);
        for (int i = 0; i < n; ++i) add(i + 1, ds[i]);
    }

    /**
     * 单点新增
     */
    void add(int i, int val) {
        while (i <= n) {
            tree[i] += val;
            i += lowbit(i);
        }
    }

    /**
     * 前缀和
     */
    int pre_sum(int i) {
        int ans = 0;
        while (i > 0) {
            ans += tree[i];
            i -= lowbit(i);
        }
        return ans;
    }
};